src/uu/truncate/src/truncate.rs

```diff
@@ -178,10 +184,27 @@ pub fn uu_app() -> Command {
 ///
 /// If the file could not be opened, or there was a problem setting the
 /// size of the file.
-fn file_truncate(filename: &str, create: bool, size: u64) -> std::io::Result<()> {
+fn file_truncate(filename: &str, create: bool, size: u64, fifo_check: FifoCheck) -> UResult<()> {
+    if fifo_check == FifoCheck::Requested {
+        if let Ok(metadata) = std::fs::metadata(filename) {
+            if metadata.file_type().is_fifo() {
+                return Err(USimpleError::new(
+                    1,
+                    format!(
+                        "cannot open {} for writing: No such device or address",
```

Comment:
This error message is indeed a clear indication that we're doing something wrong. We should be making a syscall that returns "No such device or address". That shouldn't block merging this PR though, we can indeed fix that later if you don't want to do it now.

`strace` shows this call as the one generating that error:
```
openat(AT_FDCWD, "blabla", O_WRONLY|O_NONBLOCK) = -1 ENXIO (No such device or address)
```

This Rust code:
```rust
OpenOptions::new().write(true).create(true).open("myfifo");
```
actually does the following syscall:
```
openat(AT_FDCWD, "myfifo", O_WRONLY|O_CLOEXEC)
```

Looks like the `O_NONBLOCK` is the culprit there. So maybe we can just fix this by setting the same mode as GNU via `OpenOptionsExt` and the appropriate `libc` constants?

This got me the right error message:
```rust
OpenOptions::new().write(true).custom_flags(libc::O_NONBLOCK).open("myfifo");
```

Reply:
Yes, opening the file in non-blocking mode is probably the right way forward, and your code looks correct. However, I don't want to fix this particular aspect in this PR; please note that this fifo-toctou-bug already existed before this PR, and remains unchanged.